projects/novo-elements/src/utils/app-bridge/AppBridge.spec.ts

```diff
@@ -0,0 +1,342 @@
+        }
+        return this.parent.mockSendToMe(msg, packet, this);
+    }
+
+    protected mockSendToMe(msg: MessageType, data: any, from: MockPostrobot): Promise<any> {
+        if (!(this.responders[msg])) {
+            throw new Error(`No responder for "${msg}".`);
+        }
+        const mockRobotEvent = {
+            data: {...data},
+            source: from.mockSource,
+            origin: from.mockOrigin
+        };
+        return Promise.resolve(this.responders[msg]!(mockRobotEvent))
+            .then(replyData => ({ data: {...replyData } }));
+    }
+
+    send(frameSource: Symbol, eventType: MessageType, data: any): Promise<any> {
+        if (!frameSource) {
+            throw new Error('Null frameSource on send()');
+        }
+        const recipient = MockPostrobot.allPostRobots.find(c => c.mockSource === frameSource);
+        if (recipient) {
+            return recipient.mockSendToMe(eventType, data, this);
+        } else {
+            throw new Error('Could not locate child to send to');
+        }
+    }
+}
+
+describe('AppBridge', () => {
+
+    let hostBridge: AppBridge;
+    let frame1Bridge: AppBridge;
+    let frame2Bridge: AppBridge;
+
+    let hostRobot: MockPostrobot;
+    let frame1Robot: MockPostrobot;
+    let frame2Robot: MockPostrobot;
+
+    let handleFn: jest.Mock<any, [AppBridgeHandler, any]>;
+
+    beforeEach(() => {
+
+        hostRobot = new MockPostrobot('localhost:111');
+        frame1Robot = new MockPostrobot('localhost:222', hostRobot);
+        frame2Robot = new MockPostrobot('localhost:333', frame1Robot);
+
+        // Simulate a top-level app with two nested iframes
+        hostBridge = new AppBridge('Jest Host', hostRobot);
+        hostBridge.id = 'hostapp';
+        frame1Bridge = new AppBridge('Jest Frame1', frame1Robot);
+        frame1Bridge.id = 'childapp1';
+        frame2Bridge = new AppBridge('Jest Frame2', frame2Robot);
+        frame2Bridge.id = 'childapp2';
+        hostBridge['windowOrigin'] = frame1Bridge['windowOrigin'] = frame2Bridge['windowOrigin'] = function() { return this.postRobot.mockOrigin };
+
+        handleFn = jest.fn();
+
+        // setup handlers
+        for (let handlerKey of (Object.values(AppBridgeHandler)) as AppBridgeHandler[]) {
+            hostBridge.handle(handlerKey, (event, cb) => {
+                try {
+                    return cb(handleFn(handlerKey, event));
+                } catch(err) {
+                    cb(null, err);
+                }
+            })
+        }
+    });
+
+    it('should register a new app', async () => {
+        handleFn.mockReturnValue('terry');
+        await frame1Bridge.register();
+        expect(handleFn).toHaveBeenCalledWith(AppBridgeHandler.REGISTER, jasmine.objectContaining({ id: 'childapp1' }));
+        expect(hostBridge['_registeredFrames'].length).toBe(1);
+        expect(frame1Bridge.windowName).toBe('terry');
+    });
+
+    it('opens a window', async () => {
+        // TODO
```

Comment:
is there any way we could write these unit tests instead of leaving them as TODOs? In general I am not a fan of having TODOs in the code base and I think it would be good to have unit test coverage

Reply:
Talked over this locally. While these are good tests to add, I'm not sure they're a priority just yet.